### ISSUE TITLE
Duplicate transaction hashes should be supported by the `check` tool

### ISSUE THREAD

**Issue by matheusd** [coinbase/rosetta-cli#53]
**TL;DR**

One of the tests performed by the `rosetta-cli check` tool is for the nonexistence of duplicate transaction hashes, however that is troublesome for Decred because we have a transaction reversed on-chain and then re-mined (with the exact same hash) on a different mainchain block.

Thus we need to make the check for duplicate transaction hashes either optional or remove it entirely.

**Details**

On Decred we have a mechanic that is part of our [Proof-of-Stake](https://docs.decred.org/proof-of-stake/overview/) system where voters (the PoS validators) can disapprove a specific subset of the transactions (the so called "regular transactions") contained on the block they're voting on, undoing the changes in the utxo set introduced by those transactions, effectively "reversing" them.

This always happens on the very next block of any arbitrary block. In other words, the votes on block `n+1` indicate whether the regular transactions of block `n` are approved. We only know whether block `n` is approved by the PoS validators ("voters" or "stakeholders" in Decred jargon) once block `n+1` is mined by the Proof-of-Work miners, including the necessary vote transactions.

The reversed transactions of block `n` are then eligible for re-inclusion on a future block (starting at block `n+1`).

In [our Rosetta implementation](https://github.com/decred/dcrros), we represent such transactions (txs of a block that was voted as disapproved) as operations with type `reversed` and negative amount **at block `n+1`**. That is, if a child block disapproves its parent, the **Rosetta** block `n+1` will have a copy of the regular transactions of block `n` but with a `reversed` operation and negative amounts.

This currently causes the `check` tool to fail due to the transaction hash appearing a second time on a block.

**To Reproduce**
[testnet block 460496](https://testnet.dcrdata.org/block/0000000465367ea91fc4bda606fda4e6d4040085bf61ebb68932379d07b03d89) was reversed. Thus the [c6e001](https://testnet.dcrdata.org/tx/c6e001e8e0efb435f16aa1bbadcc24ed971692aaf38e2e7922eceb7bbca5e9bb) transaction appears on our Rosetta implementation as reversed on the **next** block (block 460497):

```json
  {                                                                             
   "transaction_identifier": {                                                  
    "hash": "c6e001e8e0efb435f16aa1bbadcc24ed971692aaf38e2e7922eceb7bbca5e9bb"  
   },                                                                           
   "operations": [                                                              
    {                                                                           
     "operation_identifier": {                                                  
      "index": 0                                                                
     },                                                                         
     "type": "credit",                                                          
     "status": "reversed",                                                      
     "account": {                                                               
      "address": "TsT1FghZK4aZzo3VZMErD3rCdaC1D3XAjMJ"                          
     },                                                                         
     "amount": {                                                                
      "value": "-3347902131",                                                   
      "currency": {                                                             
       "symbol": "DCR",                                                         
       "decimals": 8                                                            
      }                                                                         
     },                                                                         
     "metadata": {                                                              
      "output_index": 0,                                                        
      "script_version": 0                                                       
     }                                                                          
    },                                              
```

**Propposed Solutions**

Either entirely disable duplicate tx detection in the check tool or make it optional so that blockchains that support multiple transactions with the same hash can be accounted for.

**Comment by patrick-ogrady** [comment]
Thanks for posting about this issue in such detail @matheusd!

The current `duplicate transaction detection` mechanism is very naive and would error on early Bitcoin coinbase transactions (prior to [BIP-34](https://github.com/bitcoin/bips/blob/master/bip-0034.mediawiki)).

Instead of making this check configurable, I think the right approach is to change the uniqueness requirement for transaction hashes to `network_identifier:block_identifier:tx_identifier`, instead of just `tx_identifier`. Many clients rely on `block_hash:tx_hash` as a globally unique identifier and making this constraint "bypassable" would lead your implementation to pass `rosetta-cli check` but would still prevent others from integrating with your implementation.
